src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/EnvelopeReferenceAppender.java

```diff
@@ -0,0 +1,91 @@
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static java.util.Arrays.asList;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Modifies case's bulkScanEnvelopes field by (conditionally) appending envelope information to existing items.
+ */
+@Service
+public class EnvelopeReferenceAppender {
+
+    private final ObjectMapper objectMapper;
+    private final Set<String> enabledServices;
+
+    public EnvelopeReferenceAppender(
+        ObjectMapper objectMapper,
+        ServiceConfiguration serviceConfiguration
+    ) {
+        this.objectMapper = objectMapper;
+
+        enabledServices = serviceConfiguration
+            .getServices()
+            .stream()
+            .filter(ServiceConfigItem::getCaseDefinitionHasEnvelopeIds)
+            .map(ServiceConfigItem::getService)
+            .collect(Collectors.toSet());
+    }
+
+    /**
+     * Appends envelope reference to the existing envelope references collection.
+     *
+     * <p>This collection represents bulkScanEnvelopes field in service case and the update will only take place
+     * if the service supports this field (i.e. has it in all its CCD case definitions).</p>
+     *
+     * @return Updated collection, if envelope info was appended. Otherwise, if the CCD case definition
+     *         for the service doesn't support envelope reference collection (bulkScanEnvelopes field),
+     *         empty Optional is returned.
+     */
+    public Optional<List<CcdCollectionElement<EnvelopeReference>>> tryAppendEnvelopeReference(
+        String service,
+        List<Map<String, Object>> existingEnvelopeReferences,
+        String envelopeId,
+        CaseAction action
+    ) {
+        if (enabledServices.contains(service)) {
+            return Optional.of(
+                appendReference(existingEnvelopeReferences, envelopeId, action)
+            );
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    private List<CcdCollectionElement<EnvelopeReference>> appendReference(
+        List<Map<String, Object>> existingEnvelopeReferences,
+        String envelopeId,
+        CaseAction action
+    ) {
+        if (existingEnvelopeReferences == null || existingEnvelopeReferences.isEmpty()) {
+            return asList(
+                new CcdCollectionElement<>(new EnvelopeReference(envelopeId, action.value))
+            );
+        } else {
+            List<EnvelopeReference> updatedReferences = existingEnvelopeReferences
+                .stream()
+                .map(ref -> objectMapper.convertValue(ref.get("value"), EnvelopeReference.class))
+                .collect(toList());
+
+            updatedReferences.add(new EnvelopeReference(envelopeId, action.value));
+
+            return updatedReferences
+                .stream()
+                .map(ref -> new CcdCollectionElement<>(ref))
+                .collect(toList());
+        }
```

Comment:
looks like an overuse of stream api. how about:

```java
var updatedReferences = existingEnvelopeReferences
    .stream()
    .map(ref -> objectMapper...)
    .map(CcdCollectionElement<>::new)
    .collect(toList());

updatedReferences.add(new CcdCollectionElement<>(new EnvelopeReference(envelopeId, action.value)));

return updatedReferences;
```

Reply:
My intention was to convert to CCD-compatible version at the very end, after the job is done. But as the job is tiny, both are fine. I'll make the change.

Reply:
done